### PR TITLE
(2204) (2205) Commenting improvements

### PR DESCRIPTION
**Pull request by CristinaRO**
## Changes in this PR
- Delivery partners can add any number of comments to an activity while there is an active report

### After
Screenshot 2021-09-27 at 18 03 54

Screenshot 2021-09-27 at 18 02 15

## Next steps

- [ ] Is an ADR required? An ADR should be added if this PR introduces a change to the architecture.
- [ ] Is a changelog entry required? An entry should always be made in `CHANGELOG.md`, unless this PR is a small tweak which has no impact outside the development team.
- [ ] Do any environment variables need amending or adding?
- [ ] Have any changes to the XML been checked with the IATI validator? See [XML Validation](https://github.com/UKGovernmentBEIS/beis-report-official-development-assistance/blob/develop/doc/xml-validation.md)

### CODE REVIEW

**Comment by CristinaRO** [comment]
[Add new comment to Activity](https://trello.com/c/ZmFCoRHH/2204-add-new-comment-to-activity)

File: app/views/staff/activities/comments.html.haml [hunk 1]
```diff
@@ -27,15 +27,29 @@
...
+                  .govuk-summary-list__row
+                    %dt.govuk-summary-list__key
+                      Date
+                    %dd.govuk-summary-list__value
+                      = I18n.l(comment.created_at.to_date)
+                  .govuk-summary-list__row
+                    %dt.govuk-summary-list__key
+                      Comment reported in
+                    %dd.govuk-summary-list__value
+                      = link_to "#{report_presenter.financial_quarter_and_year} #{report_presenter.description}", report_path(comment.report)
```
Comment (edavey) [inline app/views/staff/activities/comments.html.haml:54]: I'm not sure if we've covered these additions in an spec? Are we planning on adding some description of these additional comment properties to an existing feature spec? e.g. https://github.com/UKGovernmentBEIS/beis-report-official-development-assistance/blob/65c9142273b0c4fcfbb04bf4042c21d0bb34dca6/spec/features/staff/users_can_create_a_comment_spec.rb#L51-L60

File: app/views/staff/reports/_summary_short.html.haml [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+  .govuk-summary-list__row
+    %dt.govuk-summary-list__key
+      = t("form.label.report.deadline")
+    %dd.govuk-summary-list__value
+      = report.deadline
+  .govuk-summary-list__row
+    %dt.govuk-summary-list__key
+      = t("form.label.report.state")
+    %dd.govuk-summary-list__value
+      = report.state
```
Comment (edavey) [inline app/views/staff/reports/_summary_short.html.haml:17]: This is a useful summary to explain the context of the comment. Are you planning a feature spec to explain the benefit?

**Review by edavey — COMMENTED** [review]
Very nice work, really clear.  

I think this is still a work in progress? I've noted a couple of places where some 'feature' level detail might be helpful...

File: app/services/report/export.rb [hunk 3]
```diff
@@ -189,7 +189,7 @@ def next_quarter_forecasts
       def variance_data
         [
           variance_for_report_financial_quarter,
-          activity_presenter.comment_for_report(report_id: report_presenter.id)&.comment,
+          activity_presenter.comments_for_report(report_id: report_presenter.id).map(&:comment).join("|"),
```
Comment (edavey) [inline app/services/report/export.rb:192]: did you decide against using a `\n` in between comments?

File: spec/features/staff/users_can_create_a_comment_spec.rb [hunk 4]
```diff
@@ -4,74 +4,103 @@
...
+              expect(Comment.all.count).to eq(1)
+              expect(Comment.last.comment).to eq(form.comment)
+
+              expect(page).to have_content t("action.comment.create.success")
+
+              within ".activity_comments" do
+                expect(page).to have_content form.comment
+                expect(page).to have_content I18n.l(Date.today)
+                expect(page).to have_link "#{report.financial_quarter_and_year} #{report.description}"
+              end
```
Comment (edavey) [inline spec/features/staff/users_can_create_a_comment_spec.rb:59]: I think your "page object" approach works really well here. It's easy to follow and the fact that you are making assertions in various ways, on:

- the Capybara `form` object
- the Comment model
- the current Capybara `page`

is clear.

File: spec/features/staff/users_can_edit_a_comment_spec.rb [hunk 5]
```diff
@@ -103,21 +31,22 @@
...
-          visit organisation_activity_comments_path(activity.organisation, activity)
-          click_on t("table.body.report.edit_comment")
-          fill_in "comment[comment]", with: "Amendments have been made"
-          click_button t("default.button.submit")
-          expect(page).to have_content "Amendments have been made"
+        scenario "the user can edit any comments left by users in the same organisation" do
+          form = CommentForm.edit_from_activity_page(report: report, comment: comment)
+
+          expect(form).to have_report_summary_information
+          form.complete(comment: "Amendments have been made")
```
Comment (edavey) [inline spec/features/staff/users_can_edit_a_comment_spec.rb:38]: and the `has_report_summary_information?` matcher really pays off in communicating the high level behaviour clearly. Bravo!

I'm going to have a go at reworking those Adjustment features using this style...

**Review by edavey — APPROVED** [review]
Nice work! 

Also some really interesting feature test style to build on, taking simple helper methods to a higher level -- where they are combined with matchers and can be made chain-able.

This form object pattern was used to great effect back in https://github.com/UKGovernmentBEIS/beis-report-official-development-assistance/pull/1328

File: app/services/report/export.rb (hunk 3 above)
Comment (pezholio) [reply app/services/report/export.rb]: Oops - well spotted, I forgot that was a part of this work

File: spec/features/staff/users_can_edit_a_comment_spec.rb (hunk 5 above)
Comment (pezholio) [reply spec/features/staff/users_can_edit_a_comment_spec.rb:38]: Nice - I think there's a bunch of places we can use this approach

**Pull request merged**